src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.cs

```diff
@@ -0,0 +1,357 @@
+            _isAvailable = isAvailable;
+        }
+
+        public bool IsAvailable
+        {
+            get
+            {
+                return _isAvailable;
+            }
+        }
+    }
+
+    public class NetworkChange
+    {
+        public static event NetworkAvailabilityChangedEventHandler NetworkAvailabilityChanged
+        {
+            add
+            {
+                AvailabilityChangeListener.Start(value);
+            }
+            remove
+            {
+                AvailabilityChangeListener.Stop(value);
+            }
+        }
+
+        public static event NetworkAddressChangedEventHandler NetworkAddressChanged
+        {
+            add
+            {
+                AddressChangeListener.Start(value);
+            }
+            remove
+            {
+                AddressChangeListener.Stop(value);
+            }
+        }
+
+        internal static bool CanListenForNetworkChanges
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        internal static class AvailabilityChangeListener
+        {
+            private static object s_syncObject = new object();
+            private static Dictionary<NetworkAvailabilityChangedEventHandler, ExecutionContext> s_availabilityCallerArray =
+                new Dictionary<NetworkAvailabilityChangedEventHandler, ExecutionContext>();
+            private static NetworkAddressChangedEventHandler s_addressChange = ChangedAddress;
+            private static volatile bool s_isAvailable = false;
+            private static ContextCallback s_RunHandlerCallback = new ContextCallback(RunHandlerCallback);
+
+            private static void RunHandlerCallback(object state)
+            {
+                ((NetworkAvailabilityChangedEventHandler)state)(null, new NetworkAvailabilityEventArgs(s_isAvailable));
+            }
+
+            private static void ChangedAddress(object sender, EventArgs eventArgs)
+            {
+                lock (s_syncObject)
+                {
+                    bool isAvailableNow = SystemNetworkInterface.InternalGetIsNetworkAvailable();
+
+                    if (isAvailableNow != s_isAvailable)
+                    {
+                        s_isAvailable = isAvailableNow;
+
+                        var s_copy =
+                            new Dictionary<NetworkAvailabilityChangedEventHandler, ExecutionContext>(s_availabilityCallerArray);
+
+                        foreach (var handler in s_copy.Keys)
+                        {
+                            ExecutionContext context = s_copy[handler];
+                            if (context == null)
+                            {
+                                handler(null, new NetworkAvailabilityEventArgs(s_isAvailable));
+                            }
+                            else
+                            {
+                                ExecutionContext.Run(context.CreateCopy(), s_RunHandlerCallback, handler);
```

Comment:
We're invoking the user's callback while holding s_syncObject.  There are trivial ways that could lead to deadlock (e.g. if the callback queued and waited on a work item that tried to add a handler), though maybe we don't care about such degenerate patterns?